ccp/lib/clients/openshift/client.py

```diff
@@ -0,0 +1,258 @@
+from ccp.lib.clients.base import CmdClient
+from ccp.lib.utils.retry import retry
+from os import path
+
+
+class OpenShiftCmdClient(CmdClient):
+    """
+    Client for interacting with Openshift.
+    """
+
+    def __init__(self):
+        super(OpenShiftCmdClient, self).__init__(
+            "/usr/bin/oc"
+        )
+
+    @retry(tries=10, delay=3, backoff=2)
+    def get_sa_token_from_openshift(self, namespace, sa="sa/jenkins"):
+        """
+        Gets service account token assuming user is already logged in
+        :param namespace: The namespace of the service account.
+        :type namespace str
+        :param sa: Default sa/jenkins: The service account whose token is
+        needed
+        :type sa str
+        :return: The token, if it was able to successfully retrieve it
+        :rtype: str
+        :raises Exception
+        """
+        c1 = "%s get %s -n %s --template='{{range .secrets}}" % \
+             (self.base_command, sa, namespace)
+        c1 = c1 + "{{ .name }} {{end}}' "
+        c2 = "xargs -n 1 %s get secret -n %s --template=" \
+             % (self.base_command, namespace)
+        c2 = c2 + "'{{ if .data.token }}{{ .data.token }}{{end}}'"
+        c3 = "head -n 1"
+        c4 = "base64 -d -"
+        command = "{} | {} | {} | {}".format(
+            c1, c2, c3, c4
+        )
+        return self.run_command(command, shell=True)
+
+    @retry(tries=5, delay=3, backoff=3)
+    def get_token_from_mounted_secret(
+            self, secret_mount_path="/tmp/jenkins-secret"
+    ):
+        """
+        Get token from a mounted secret.
+        :param secret_mount_path: The path where the secret is mounted
+        :return: The token, if it was able to retrieve it.
+        """
+        token_location = secret_mount_path + "/token"
+        if not path.exists(token_location):
+            raise Exception(
+                "Could not find token at {}".format(
```

Comment:
we are raising all the exceptions in the function, lets make sure we are actually caching them properly. Program should not stop working as expected, even though there is exception

Reply:
Yes.  But it should be handled at higher level the way they want to handle it.

Reply:
FYI exceptions should rarely be handled at lower level of code. Unless there is explicit need. This should be done at higher level as each client of lib code will want to handle exceptions in different way. Not even processor. 

What will end up happening is API server for example will be the highest level that needs to know there is exception and handle it correctly.

Another example will be index reader. It should be the one that decides what action it wants to take it when an oc command fails for example. The duty of lower level code that runs the command is updating the reader of the condition.